src/service/performance-impl.js

```diff
@@ -269,18 +270,35 @@ export class Performance {
       return Promise.resolve();
     }
 
-    return channelPromise.then(() => {
-      this.isMessagingReady_ = true;
+    return channelPromise
+      .then(() => {
+        // Tick the "messaging ready" signal.
+        this.tickDelta('msr', this.win.Date.now() - this.initTime_);
 
-      // Tick the "messaging ready" signal.
-      this.tickDelta('msr', this.win.Date.now() - this.initTime_);
+        return this.maybeAddStoryExperimentId_();
+      })
+      .then(() => {
+        this.isMessagingReady_ = true;
 
-      // Forward all queued ticks to the viewer since messaging
-      // is now ready.
-      this.flushQueuedTicks_();
+        // Forward all queued ticks to the viewer since messaging
+        // is now ready.
+        this.flushQueuedTicks_();
 
-      // Send all csi ticks through.
-      this.flush();
+        // Send all csi ticks through.
+        this.flush();
+      });
+  }
+
+  /**
+   * Add a story experiment ID in order to slice the data for amp-story.
+   * @return {!Promise}
+   * @private
+   */
+  maybeAddStoryExperimentId_() {
+    return isStoryDocument(getAmpdoc(this.win.document)).then(isStory => {
```

Comment:
Looks like `isStoryDocument` checks the doc body. You mentioned waiting for that might introduce latency. Is this no longer a concern?

Reply:
I've been able to work around this concern thanks to the architecture of the performance service. It has a queuing system, where it stores all the metrics it receives until it's ready to communicate to the viewer.
This implementation makes it wait for the body to be parsed before emptying the queued metrics, which I think should be fine.

It ensures the `story` experiment is added to the list before they're passed to the viewer with the first call to `sendCSI`.